File: app/src/main/java/com/example/dkdus/cashbook/AddActivity.java
package com.example.dkdus.cashbook;

import androidx.annotation.RequiresApi;
import androidx.appcompat.app.AppCompatActivity;
import androidx.databinding.DataBindingUtil;

import android.app.DatePickerDialog;
import android.content.Intent;
import android.os.Build;
import android.os.Bundle;
import android.util.Log;
import android.view.View;
import android.widget.DatePicker;
import android.widget.RadioButton;
import android.widget.TextView;

import com.example.dkdus.cashbook.databinding.ActivityAddBinding;
import com.example.dkdus.cashbook.model.ChildList;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.auth.FirebaseUser;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;

import java.time.LocalDate;
import java.time.LocalTime;
import java.time.format.DateTimeFormatter;
import java.util.HashMap;
import java.util.Map;

@RequiresApi(api = Build.VERSION_CODES.O)
public class AddActivity extends AppCompatActivity {

    private static final String TAG = AddActivity.class.getSimpleName();
    private FirebaseDatabase firebaseDatabase = FirebaseDatabase.getInstance();
    private DatabaseReference databaseReference = firebaseDatabase.getReference();
    DateTimeFormatter Dateformatter = DateTimeFormatter.ofPattern("yyyy년 MM월 dd일");
    DateTimeFormatter Hourformatter = DateTimeFormatter.ofPattern("HHmmss");
    String today_date;
    private FirebaseAuth mAuth;
    private FirebaseUser mUser;
    private String time;
    private TextView textDate;
    private boolean update = false;
    private String intentdate;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        final ActivityAddBinding binding= DataBindingUtil.setContentView(this,R.layout.activity_add);

        today_date = LocalDate.now().format(Dateformatter);
        binding.inputDate.setText(today_date);
        time = LocalTime.now().format(Hourformatter);
        Intent intent = getIntent();

        if(intent != null && (intent.getStringExtra("date"))!= null){
            update = true;
            String intenttype = intent.getStringExtra("type");
            String intentcontents = intent.getStringExtra("contents");
            String intentmoney = intent.getStringExtra("money");
            intentdate = intent.getStringExtra("date");
            String intentcategory = intent.getStringExtra("category");
            time = intent.getStringExtra("time");

            int id = whatcategory(intentcategory);
            RadioButton radioButton = (RadioButton) findViewById(id);
            radioButton.setChecked(true);
            binding.inputType.setText(intenttype);
            binding.inputContents.setText(intentcontents);
            binding.inputMoney.setText(intentmoney);
            binding.inputDate.setText(intentdate);
        }

        mAuth = FirebaseAuth.getInstance();
        mUser = mAuth.getCurrentUser();

        binding.inputDate.setOnClickListener(new View.OnClickListener() {
            LocalDate todayDate;
            @RequiresApi(api = Build.VERSION_CODES.O)
            @Override
            public void onClick(View v) {
                todayDate = LocalDate.parse(binding.inputDate.getText(), Dateformatter);
                DatePickerDialog datePickerDialog = new DatePickerDialog(AddActivity.this, dateSetListener,
                        todayDate.getYear(),todayDate.getMonthValue(),todayDate.getDayOfMonth());
                binding.inputDate.setText(today_date);
                datePickerDialog.show();
            }
        });

        final String finalTime = time;
        binding.inputButton.setOnClickListener(new View.OnClickListener() {

            @RequiresApi(api = Build.VERSION_CODES.O)
            @Override
            public void onClick(View v) {

                String IDdate = binding.inputDate.getText().toString();

                int id = binding.radioGroup.getCheckedRadioButtonId();
                RadioButton radioButton = (RadioButton) findViewById(id);
                String category = radioButton.getText().toString();

                String type = binding.inputType.getText().toString();
                String contents = binding.inputContents.getText().toString();
                String money = binding.inputMoney.getText().toString();

                ChildList childList = new ChildList(type, contents, money, IDdate, category, finalTime);

                Map<String, Object> childUpdates = new HashMap<>();

               childUpdates.put("/users"+"/"+mUser.getUid()+"/"+IDdate+"/"+ finalTime, childList);
               databaseReference.updateChildren(childUpdates);

               //날짜를 변경할 경우 처리
               if(update && !IDdate.equals(intentdate)){
                   DatabaseReference reference = firebaseDatabase.getReference().child("users").child(mUser.getUid()).child(intentdate).child(time);
                   reference.removeValue();
               }
                Intent gointent = new Intent(AddActivity.this, MainActivity.class);
               startActivity(gointent);
            }
        });
    }

    private int whatcategory(String intentcategory) {
        if(intentcategory.equals("지출")){
            return R.id.radioButton1;
        } else if(intentcategory.equals("수입")){
            return R.id.radioButton2;
        } else if(intentcategory.equals("저축")){
            return R.id.radioButton3;
        }

        return 0;
    }

    private DatePickerDialog.OnDateSetListener dateSetListener = new DatePickerDialog.OnDateSetListener(){
        @RequiresApi(api = Build.VERSION_CODES.O)
        @Override
        public void onDateSet(DatePicker view, int year, int month, int dayOfMonth) {
            LocalDate today = LocalDate.of(year, month, dayOfMonth);
            today_date = today.format(Dateformatter);
           // textDate.setText(today.format(Dateformatter));
        }
    };
}

File: app/src/main/java/com/example/dkdus/cashbook/activity/AddActivity.kt
package com.example.dkdus.cashbook.activity

import android.app.DatePickerDialog
import android.app.DatePickerDialog.OnDateSetListener
import android.content.Intent
import android.os.Build
import android.os.Bundle
import android.view.View
import android.widget.RadioButton
import android.widget.TextView
import androidx.annotation.RequiresApi
import androidx.appcompat.app.AppCompatActivity
import androidx.databinding.DataBindingUtil
import com.example.dkdus.cashbook.R
import com.example.dkdus.cashbook.databinding.ActivityAddBinding
import com.example.dkdus.cashbook.model.ChildList
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.auth.FirebaseUser
import com.google.firebase.database.FirebaseDatabase
import kotlinx.android.synthetic.main.activity_add.*
import java.time.LocalDate
import java.time.LocalTime
import java.time.format.DateTimeFormatter
import java.util.*

@RequiresApi(api = Build.VERSION_CODES.O)
class AddActivity : AppCompatActivity() {
    private val firebaseDatabase = FirebaseDatabase.getInstance()
    private val databaseReference = firebaseDatabase.reference
    var Dateformatter = DateTimeFormatter.ofPattern("yyyy년 MM월 dd일")
    var Hourformatter = DateTimeFormatter.ofPattern("HHmmss")
    var todayDate: String? = null
    private lateinit var mAuth: FirebaseAuth
    private lateinit var mUser: FirebaseUser
    private lateinit var time: String
    private lateinit var date : String
    private var update = false

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_add)
//        val binding: ActivityAddBinding = DataBindingUtil.setContentView(this, R.layout.activity_add)
        todayDate = LocalDate.now().format(Dateformatter)
        input_date.setText(todayDate.toString())
        time = LocalTime.now().format(Hourformatter)

        if (intent is Intent){
            var childList : ChildList? = intent.getParcelableExtra("list")
            if(childList is ChildList){

                update = true
                time = childList.time!!
                date = childList.date!!

                val id = whatCategory(childList.category!!)
                val radioButton = findViewById<View>(id) as RadioButton

                radioButton.isChecked = true
                input_type.setText(childList.type)
                input_contents.setText(childList.contents)
                input_money.setText(childList.money)
                input_date.setText(childList.date)
                }
            }
        mAuth = FirebaseAuth.getInstance()
        mUser = mAuth.currentUser!!
        input_date.setOnClickListener(object : View.OnClickListener {

            @RequiresApi(api = Build.VERSION_CODES.O)
            override fun onClick(v: View) {
                val todayDate = LocalDate.parse(input_date.text, Dateformatter)
                var datePickerDialog = DatePickerDialog(this@AddActivity, dateSetListener,
                        todayDate.getYear(), todayDate.getMonthValue(), todayDate.getDayOfMonth())
                input_date.setText(todayDate.toString())
                datePickerDialog.show()
            }
        })
        val finalTime = time
        input_button.setOnClickListener {
            val IDdate = input_date.text.toString()
            val id = radioGroup.checkedRadioButtonId
            val radioButton = findViewById<View>(id) as RadioButton
            val category = radioButton.text.toString()
            val type = input_type.text.toString()
            val contents = input_contents.text.toString()
            val money = input_money.text.toString()

            val childList = ChildList(type, contents, money, IDdate, category, finalTime)
            val childUpdates: MutableMap<String, Any> = HashMap()
            childUpdates["/users" + "/" + mUser!!.uid + "/" + IDdate + "/" + finalTime] = childList
            databaseReference.updateChildren(childUpdates)

            //날짜를 변경할 경우 처리
            if (update && IDdate != date) {
                val reference = firebaseDatabase.reference.child("users").child(mUser.uid).child(date).child(time)
                reference.removeValue()
            }
            val goIntent = Intent(this@AddActivity, MainActivity::class.java)
            startActivity(goIntent)
        }
    }

    private fun whatCategory(intentCategory: String): Int {
        if (intentCategory == "지출") {
            return R.id.radioButton1
        } else if (intentCategory == "수입") {
            return R.id.radioButton2
        } else if (intentCategory == "저축") {
            return R.id.radioButton3
        }
        return 0
    }

    private val dateSetListener = OnDateSetListener { view, year, month, dayOfMonth ->
        val today = LocalDate.of(year, month, dayOfMonth)
        todayDate = today.format(Dateformatter)
    }

    companion object {
        private val TAG = AddActivity::class.java.simpleName
    }
}